### ISSUE TITLE
GroupBatch

### ISSUE THREAD

**Issue by Arithmomaniac** [morelinq/MoreLINQ#341]
A proposed cross between `Batch` and `GroupBy` which I have used internally in some projects.

```csharp
IEnumerable<IGrouping<TKey, TSource>> GroupBatch<TKey, TSource>(
    this IEnumerable<TSource> source, Func<TSource, TKey> keySelector, int size)
```

Basically, it works like GroupBy, but instead of needing to go through the whole source input to build the object (which may not be possible), it yields `IGrouping`s with at most a fixed number of items (so multiple groupings may have the same key). Any leftover items are yielded in order of when they were created.

For example, `Enumerable.Range(0,6).GroupBatch(IsOdd, 2)` would return the groupings:

 * `false : [0, 2]`
 * `true : [1, 3]`
 * `false : [4]`
 * `true : [5]`

The internal implementation is inspired by `Batch`, but a `Dictionary` is used instead of a `List`. I have a naive implementation in this [Gist](https://gist.github.com/Arithmomaniac/2a36321ae2e613b87f84a5749ef88c82), and can turn it (after some work) into a full pull request.

**Comment by Arithmomaniac** [comment]
You would first get `[1,1]`, because that is the first item to hit the batch size limit.

Then you would get the leftover/incomplete batches in the order _that current batch_ was created. (That was just what I chose, other options are fine.) so `[2]`, `[3]`, `[1]`.

My usual use case is something like the following:
 * Stream tens of thousands of items from an API (too many to fit into memory)
 * Apply `GroupBatch` to get batches of 1000 items
 * Pass the batch grouping values into another API, where the endpoint depends on the key of the batch grouping.

I think of it as a way to "split" an input stream on a single thread.

**Comment by atifaziz** [comment]
@Arithmomaniac What's the difference between or benefit over batching first and then grouping each batch? So you batch a thousand items at a time and then create groups from those thousand items.

**Comment by Arithmomaniac** [comment]
Because then you don't have 1000 items at a time in those batches; you'd have, say, 350, 400, and 250. This gives a predictable batch size.

**Comment by fsateler** [comment]
> Then you would get the leftover/incomplete batches in the order that current batch was created. (That was just what I chose, other options are fine.) so [2], [3], [1].

I think you misunderstood me. The source sequence in my example is infinite (`new [] { 1 }.Repeat()` is an infinite list of `1`). That's why I ask if 2 and 3 will ever be yielded.

**Comment by Arithmomaniac** [comment]
Missed that. Sorry.
No they would not.

-----

**Comment by atifaziz** [comment]
> This gives a predictable batch size.

@Arithmomaniac I get that but I'm trying to understand why having a predictable size of groups helping any more the memory issue listed in your use case than batching first and then grouping (yielding potentially uneven groups). Is it because doing work on very small groups, such as submitting them to a remote endpoint, would be wasteful?

**Comment by Arithmomaniac** [comment]
> Is it because doing work on very small groups, such as submitting them to a remote endpoint, would be wasteful?

@atifaziz correct. In such a case. this is more memory-efficient than the alternative of getting even-sized groups via grouping-then-batching.

**Comment by Arithmomaniac** [comment]
As a side note, I think `BatchBy` or `BatchedGroupBy` may be more explanatory method names.
